src/state/mod.rs

```diff
@@ -37,75 +16,128 @@ pub struct State {
+        // Texture's bindgroup
+        let diffuse_bind_group = Texture::init_texture_bind_group(
+            &device,
+            &diffuse_texture,
+            &texture_bind_group_layout,
+            "diffuse_texture_bind_group",
+        );
+
+        // Shaders
+        let shader = device.create_shader_module(wgpu::include_wgsl!("../shaders/shader.wgsl"));
 
-        // END NEW
+        // Buffers
+        let vertex_buffer = Buffer::new(
+            &device,
+            Vertex::VERTICES_PENTAGON,
+            wgpu::BufferUsages::VERTEX,
+            "Vertex Buffer",
+        );
+        let index_buffer = Some(Buffer::new(
+            &device,
+            Vertex::INDICES,
+            wgpu::BufferUsages::INDEX,
+            "Index Buffer",
+        ));
 
-        let render_pipeline = RenderPipeline::new(&device, &config, &[&texture_bind_group_layout]);
+        // Camera
+        let camera = Camera::new(&config);
+        let camera_controller = CameraController::new(0.2);
+        // Camera's Matrix
+        let mut camera_uniform = CameraUniform::new();
+        camera_uniform.update_view_proj(&camera);
+        // Camera's buffer
+        let camera_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+            label: Some("Camera Buffer"),
+            contents: bytemuck::cast_slice(&[camera_uniform]),
+            usage: wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST,
+        });
+        // Camera's bindgroup layout
+        let camera_bind_group_layout =
+            device.create_bind_group_layout(&Camera::get_camera_bindgroup_layout_descriptor());
+        // Camera's bindgroup
+        let camera_bind_group = Camera::init_camera_bind_group(
+            &device,
+            &camera_bind_group_layout,
+            "camera_bind_group",
+            &camera_buffer,
+        );
 
-        let clear_color = wgpu::Color::BLACK;
+        // Render Pipeline
+        let render_pipeline = Self::init_render_pipeline(
+            &device,
+            &config,
+            &shader,
+            "vs_main",
+            "fs_main",
+            &[&texture_bind_group_layout, &camera_bind_group_layout],
+        );
 
-        #[rustfmt::skip]
-        let vertex_buffer = Buffer::new(&device, VERTICES, wgpu::BufferUsages::VERTEX, "Vertex buffer");
-        #[rustfmt::skip]
-        let index_buffer = Some(Buffer::new(&device, INDICES, wgpu::BufferUsages::INDEX, "Index Buffer"));
 
         Self {
             window,
             surface,
             device,
             queue,
             config,
-            size,
-            clear_color,
+            size: window_size,
+            clear_color: wgpu::Color::BLACK,
             render_pipeline,
             vertex_buffer,
             index_buffer,
-            diffuse_bind_group,
-            diffuse_texture,
+            num_indices,
+            bool_event,
```

Comment:
usually, I think it's not a good idea to name something `bool_event` :v, because we only know the type of this value, but not what it indicates. However, the type encoded in the name is not needed in static typing language, because we always have compiler with us, right?

In my opinion, the name should reflect what it is. In this case, maybe it's input_key_pressed for example

Reply:
the `bool_event` will be refactored or replaced in the future. Its purpose is only for the challenge 3 in the tutorial.